### ISSUE TITLE
Add module_set_weight() API function

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#4073]
**Description of the need**
Right now when we need to set module weight (which determines hook execution order), a module has to manually update its weight in the system table in hook_install() like this:

```
/**
 * Implements hook_install().
 */
function contextual_install() {
  // Weight must be greater than that of views module.
  db_query("UPDATE {system} SET weight = 11 WHERE name = 'contextual'");
}
```

**Proposed solution**
Drupal 8 provided a [module_set_weight()](https://api.drupal.org/api/drupal/core%21includes%21module.inc/function/module_set_weight/8.2.x) function. Though ironically it's intended to be deprecated and possibly removed.

**Alternatives that have been considered**
We could just keep using SQL queries, but providing an API function could allow us to more easily switch the system table to a config file later.

**Comment by BWPanda** [comment]
Ooh, yes please! This might also provide automatic documentation on the process for doing this (I was just searching for best practices regarding setting a module's weight but couldn't find anything on the API site).

**Comment by BWPanda** [comment]
Here's a PR: https://github.com/backdrop/backdrop/pull/3041
I imagine this'll need a test or two, but I'm not sure how best to do that, so hopefully someone else can use this as a start...

**Comment by herbdool** [comment]
Does this need a test? Do other utility functions here have a test?

**Comment by klonos** [comment]
Should we name the function `project_set_weight()` so to include themes/layouts/profiles in case these also need specifying their weight?

**Comment by herbdool** [comment]
@klonos I would say no. I believe weights don't make sense for themes or profiles. You can't have multiple profiles operating at the same time. And I'm pretty sure that's true of themes too. The weight is set to zero for all enabled and disabled themes.

**Comment by indigoxela** [comment]
Wouldn't it make sense if the core modules currently using the SQL syntax implemented the hook instead?

Listing:
```
> grep -nr 'UPDATE {system} SET weight' core

core/modules/contextual/contextual.install:12:  db_query("UPDATE {system} SET weight = 11 WHERE name = 'contextual'");
core/modules/contextual/contextual.install:24:  db_query("UPDATE {system} SET weight = 11 WHERE name = 'contextual'");
core/modules/views/views.install:11:  db_query("UPDATE {system} SET weight = 10 WHERE name = 'views'");
core/modules/system/system.install:2588:  db_query("UPDATE {system} SET weight = 60 WHERE name = 'layout' AND type = 'module'");
core/modules/layout/layout.install:35:  db_query("UPDATE {system} SET weight = 60 WHERE name = 'layout' AND type = 'module'");
```

**Comment by BWPanda** [comment]
I wasn't sure if it needed tests, and @herbdool questioned the need for tests, so I'd say let's leave that tag off until someone definitively states that this does need tests...

**Comment by BWPanda** [comment]
PR updated with @indigoxela's suggestion of actually using the function in core :smiley:

**Comment by indigoxela** [comment]
I wonder how we can properly test that...

The first thing might be to restart the sandbox to check if everything still works in a fresh install. (The 5.3 testbox seems broken (404) anyway.)

**Comment by BWPanda** [comment]
See how that goes...

**Comment by indigoxela** [comment]
The install worked fine and all tests were passing.

To test this locally, where I can actually take a look at database entries, I applied the changes. (Which was a bit hard because the PR branch lags behind.)

1. I uninstalled the contextual module
2. Removed the db entry in the system table manually - just to make sure
3. Installed contextual module again
4. Checked the new entry in the system table - weight was "11" again

Hope this test makes sense. :wink:

**Comment by quicksketch** [comment]
The PR looks good to me. The only thing that gives me pause is that I don't think that there should be a default of `0` for the `$weight` parameter. I don't think a default makes sense here when you will always want to be explicit about the weight that is being set. In the PR there is one line using the default:

```
module_set_weight('block');
```

But the default just makes me confused, what is the weight value being set? You have to look up `module_set_weight()` to find out the default is going to be zero.

**Comment by BWPanda** [comment]
Good point @quicksketch, PR updated to remove default value.

**Comment by quicksketch** [comment]
Thanks @BWPanda!

I made a mistake here... we probably should have only merged this into 1.x and this should have been an API addition for 1.18.0. However, it won't break anything and contrib modules can continue using the old approach to maintain compatibility with older versions of Backdrop.

I also added a `@since 1.17.2 Function added.` docblock to help identify this function as a new feature.
